### DISCUSSION TITLE
Add New "Phase Offset" Field to Oscillator Component

### DISCUSSION THREAD

**Discussion by austephner** [FakeFishGames/Barotrauma#8872] [Ideas and suggestions]
My suggestion/wish for the oscillator component:
1. Add new public float variable called `phaseOffset`
2. Value should be used alongside all oscillator component wave calculations
3. Expose `phaseOffset` in game with a new wiring input node as `set phase offset`

Found it kind of weird we can't adjust the phase of the oscillator or at least "starting phase" or "phase offset". In some cases I don't want my oscillators to be running on the same phase. 

https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/OscillatorComponent.cs

**Comment by mygamingaccount** [comment]
Isn't this the same as adding a Delay component?

**Reply by austephner** [reply]
looks like adding a delay after the oscillator does behave as expected. I guess that's a good way of offsetting phase. 

There aren't too many drawbacks to this other than delayed start of the oscillator's output for whatever I'm using it for. I'd rather start the oscillators all at the same time and have their output already adjusted/available

**Reply by mygamingaccount** [reply]
I agree that the two aren't identical. Does the glitch at t=0 cause any issues though?

**Reply by austephner** [reply]
apologies, not sure what you mean by "the glitch at t=0"!

**Reply by mygamingaccount** [reply]
Consider three triangle wave generators 120 degrees apart.
If you use time delay, then at t=0 two outputs will be zero, with the second phase coming on at 120˚, and the third at 240˚.
This is different from having a phase delay, since the first time the sub loads you will have the correct signals then.

I do not know where this difference would be important.

**Reply by austephner** [reply]
Ah, I see what you meant now. Yes, the difference is negligible in probably 99% of use cases, which is likely to put a nail in the coffin of this thread. However, I did find another major drawback.

I perform a calculation that's meant to adjust phase real time. The delay components don't seem to like this. I looked through the code, we can see that whenever delay is set the entire signal queue is cleared if the incoming value doesn't match the current delay value (line 128):

https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/DelayComponent.cs

This results in my components' outputs stuttering towards the expected target value as it seems like sometimes the signal queue is reset before the delay can fire off again. The phase coincides with the sub's velocity and thus changes quite often. Phase has to be recalculated to match the oscillator's frequency correctly. So the resulting effect doesn't work very well with constantly changing numbers, I actually had to rip that part out of my sub.

Is it a big deal? Nope, this is just a suggestion thread :D

**Reply by mygamingaccount** [reply]
In the remaining 1% you can make your own oscillator with a few components, which will have a phase input :)

Two MEM and a adder to make a ramp function, divide with a constant to set the frequency, sine component for the shape. Another adder between the divider output and the sine component will set the phase.

**Reply by austephner** [reply]
hmm ... yeah that works nicely. I can see how this is viable option. In the below example I didn't do an additional adder/mem for phase, but it would be an easy insert. I'm still wrapping my head around how to take high level programming concepts and put them into barotrauma's component system!

![image](https://user-images.githubusercontent.com/5714722/165341817-f941ed26-aca6-4859-90d5-2a583d0f00eb.png)

Thanks for the suggestion/workaround.

**Reply by mygamingaccount** [reply]
That's right, and from the same ramp generator you can have multiple adders and sine generators so the base frequency will be linked and the phase offset will remain consttant
